src/connectors/infinite-hits/connectInfiniteHits.ts

```diff
@@ -207,92 +207,110 @@ const connectInfiniteHits: InfiniteHitsConnector = function connectInfiniteHits(
-        // hits widgets mounted on the page.
-        (results.hits as any).__escaped = initialEscaped;
+        const widgetRenderState = this.getWidgetRenderState(renderOptions);
 
         const cachedHits = cache.read({ state }) || {};
-        if (cachedHits[page] === undefined) {
-          cachedHits[page] = results.hits;
-          cache.write({ state, hits: cachedHits });
-        }
-
-        const firstReceivedPage = getFirstReceivedPage(state, cachedHits);
-        const lastReceivedPage = getLastReceivedPage(state, cachedHits);
-        const isFirstPage = firstReceivedPage === 0;
-        const isLastPage = results.nbPages <= lastReceivedPage + 1;
+        const { page = 0 } = state;
 
         sendEvent('view', cachedHits[page]);
 
         renderFn(
           {
-            hits: extractHitsFromCachedHits(cachedHits),
-            results,
-            sendEvent,
-            bindEvent,
-            showPrevious,
-            showMore,
-            isFirstPage,
-            isLastPage,
+            ...widgetRenderState,
             instantSearchInstance,
-            widgetParams,
           },
           false
         );
       },
 
+      getRenderState(renderState, renderOptions) {
+        return {
+          ...renderState,
+          infiniteHits: this.getWidgetRenderState(renderOptions),
+        };
+      },
+
+      getWidgetRenderState({ results, helper, state, instantSearchInstance }) {
+        let isFirstPage;
+        const cachedHits = cache.read({ state }) || {};
+
+        if (!results) {
+          showPrevious = getShowPrevious(helper, cachedHits);
+          showMore = getShowMore(helper, cachedHits);
+          sendEvent = createSendEventForHits({
+            instantSearchInstance,
+            index: helper.getIndex(),
+            widgetType: this.$$type!,
+          });
+          bindEvent = createBindEventForHits({
+            index: helper.getIndex(),
+            widgetType: this.$$type!,
+          });
+          isFirstPage =
+            helper.state.page === undefined ||
+            getFirstReceivedPage(helper.state, cachedHits) === 0;
+        } else {
+          const { page = 0 } = state;
+
+          if (escapeHTML && results.hits.length > 0) {
+            results.hits = escapeHits(results.hits);
+          }
+          const initialEscaped = (results.hits as any).__escaped;
+
+          results.hits = addAbsolutePosition(
+            results.hits,
+            results.page,
+            results.hitsPerPage
+          );
+
+          results.hits = addQueryID(results.hits, results.queryID);
+
+          results.hits = transformItems(results.hits);
+
+          // Make sure the escaped tag stays after mapping over the hits.
+          // This prevents the hits from being double-escaped if there are multiple
+          // hits widgets mounted on the page.
+          (results.hits as any).__escaped = initialEscaped;
+
+          if (cachedHits[page] === undefined) {
```

Comment:
the cache read should probably move from render to before here, wdyt?

Reply:
I did not understood what you meant by that, can you explain more precisely?

(all this side effects like update of the `results` object and `cache.write` should not be in `getWidgetRenderState` which should be a pure getter but I did not wanted to mess with the logic here for now)

Reply:
As @Haroenv and @francoischalifour discussed in another PR, when we get rid of `init` in the future, it's natural that `getWidgetRenderState` is the place where side effects occur.

Although `getWidgetRenderState` has side effects, I think it's okay as long as it's idempotent. In this case, it's okay to call `getWidgetRenderState` multiple times as long as it reads the cache and writes the same one (not changing the cache). And, of course, `getWidgetRenderState` shouldn't trigger `sendEvent` inside it.

To sum up, I think it's good to have `cache.read` inside `getWidgetRenderState`. And we add another property to `InfiniteHitsWidgetRenderState`, which is something like `currentPageHits`. If so, inside `render` we can get `currentPageHits` from the render state and call `sendEvent`.

What do you think?